dask/dataframe/io/parquet/core.py

```diff
@@ -184,15 +185,35 @@ def read_parquet(
     read_from_paths : bool, default None
         Only used by ``ArrowDatasetEngine`` when ``filters`` are specified.
         Determines whether the engine should avoid inserting large pyarrow
-        (``ParquetFileFragment``) objects in the task graph. If this option
-        is True (default), ``read_partition`` will need to regenerate the
-        appropriate fragment object from the path and row-group IDs.  This will
-        reduce the size of the task graph, but will add minor overhead to
-        ``read_partition``.
-    chunksize : int, str
-        The target task partition size.  If set, consecutive row-groups
-        from the same file will be aggregated into the same output
-        partition until the aggregate size reaches this value.
+        (``ParquetFileFragment``) objects in the task graph.  If this option
+        is True, ``read_partition`` will need to regenerate the appropriate
+        fragment object from the path and row-group IDs.  This will reduce the
+        size of the task graph, but will add minor overhead to ``read_partition``.
+        By default (None), ``ArrowDatasetEngine`` will set this option to
+        ``False`` when there are filters.
+    chunksize : int, str, tuple, default None
+        The target output partition size. If specified, adjacent row-groups
+        and/or files will be aggregated into the same output partition until
+        the cumulative parquet-storage size reaches this value. By default,
+        inter-file aggregation will only occur if said files are stored in
+        the same directory.
+
+        Note that a non-default chunksize-aggregation range can be specified
+        by passing ``chunksize`` as a tuple with two elements. The first
+        element should be the usual int or str-based representation of the
+        desired partition size. The second element should be a string
+        specifying the aggregation range. The value "file" declares that
+        aggregation may only occur within a single file. The value "directory"
+        declares that aggregation may only occur within an immediate directory.
+        If a different string value is specified it will be used for regular
+        expression matching in the file path. For example, when reading a dataset
+        partitioned by multiple columns, like "month" and "day", we can set
+        the aggregation range to be the "day" partition by passing something
+        like the following for ``chunksize``::
+
+            ``chunksize=("1MiB", "day=[^/]*/")``
+
+        Note that the example above assumes a "/" directory separator.
```

Comment:
The other comment was already resolved so I copied my comment for it to not get lost.

I think allowing for "non-default" aggregations across directories is a very useful feature!
Is there an actual usecase for allowing regex? In my experience, most of the aggregations users would care about are partition columns / directories or "free-for-all". The regex might be very powerful but it also gives the wrong incentive imho. I don't think we should steer people into encoding information in the file paths. In my experience it is almost always easier and more performant to encode this information in the parquet files themselves
Any reason why the "aggregation level" / second tuple element is not an argument of its own? It is not necessarily a "chunk size", is it?

For reference, we have something similar in kartothek which is called dispatch_by (see parameters of https://kartothek.readthedocs.io/en/latest/_rst/kartothek.io.dask.dataframe.html#kartothek.io.dask.dataframe.read_dataset_as_ddf) which concatenates entire files based on the given index columns.

Another idea: Allow aggregations for all row groups with overlapping statistics in a given column, regardless of where the RG comes from? Out of scope for this PR but maybe something to try out / play with? idk

Reply:
>Is there an actual usecase for allowing regex? In my experience, most of the aggregations users would care about are partition columns / directories or "free-for-all". The regex might be very powerful but it also gives the wrong incentive imho. I don't think we should steer people into encoding information in the file paths.

This concern makes sense -  It does seem possible that the best approach to file-aggregation is to leverage the existing directory/hive-partitioning concept if that does happen to cover the dominant "directory-group" `chunksize` use case.  The only issue is that partition discovery and handling is all very "engine specific" at the moment, so we would need further changes in each of the engines to add something like a general partition label to each row-group statistic.  Such an approach is fine, but it wouldn't capture the case that a subdirectory does not actually correspond to a known dataset column (which may be totally fine.. I'm really not sure).  @randerzander - Do you have any use cases where the directory group is *not* a column partition?

Overall, the approach in this PR is to use the limited information in the existing `part` statistics to aggregate adjacent row-groups/files.  Expression matching is an *easy* way to enable "general" directory-group aggregation without requring more information from the engine, but I agree that it may not be the *best* way to achieve what we need.

>Any reason why the "aggregation level" / second tuple element is not an argument of its own? It is not necessarily a "chunk size", is it?

No good reason besides an effort to avoid adding more parameters.  The aggregation range is certainly not the chunksize, but it is only relevant when chunksize is used anyway.

Reply:
> Such an approach is fine, but it wouldn't capture the case that a subdirectory does not actually correspond to a known dataset column 

This is a question about compatibility. What kind of "dataset formats" do we want to support? My impression, so far, was that we support only "hive-like" dataset formats with a few special cases. Hive-like dataset formats in this context means a key structure following a filesystem-like directory structure where the directories are named after fields of the "table" the dataset represents. What I call "dataset format" here is probably best caught by the "engine" abstraction here. When I'm talking about "dataset format" my head thinks of Apache Iceberg, Delta Lake, Arrow Dataset, fastparquet?(actually I don't know much about fastparquet), kartothek, plain Hive, etc.

In fact, the special cases we support are things like arrow dataset where we build on top of the hive-like structure by providing `_common_metadata` and sometimes `_metadata`.

If this is all we support we can put this assumption in the code and be sometimes a bit more efficient. However, I haven't followed the most recent developments, especially not in arrow dataset so I might have the wrong impression on some of the points here.

What I'm trying to say is that I believe we should try to be more explicit in what we consider "compatible" and what we do not support. Requiring people to adhere to some standard is not necessarily a bad thing.

For example, according to our own docs 

> Read a Parquet file into a Dask DataFrame
>
> This reads a directory of Parquet data into a Dask.dataframe, one file per partition.

something like the following _might_ be supported, depending on how loosely you interpret "a directory of data". 

```
dataset_prefix/data.parquet
dataset_prefix/some_directory/data.parquet
dataset_prefix/another_directory/some_directory/data.parquet
```

but I would be very hesitant to support this unstructured collection of files, at least not with the full feature list.

I realize this opens a discussion which is **much larger** than the current PR so feel free to not discuss this in detail here. I suggest to jump on a call sometime to align expectations here and discuss where we're headed.

Reply:
>What kind of "dataset formats" do we want to support? My impression, so far, was that we support only "hive-like" dataset formats with a few special cases. Hive-like dataset formats in this context means a key structure following a filesystem-like directory structure where the directories are named after fields of the "table" the dataset represents. What I call "dataset format" here is probably best caught by the "engine" abstraction here. When I'm talking about "dataset format" my head thinks of Apache Iceberg, Delta Lake, Arrow Dataset, fastparquet?(actually I don't know much about fastparquet), kartothek, plain Hive, etc.

Your understanding is correct, and I really appreciate that you are taking the time to discuss this.

Dask supports four general cases: (1) A single file, (2) a single directory containing an arbitrary number of files, (3) a "hive-like" dataset format, and (4) an arbitrary list of files.  I completely agree that it is much easier to design and maintain an API if we are explicit about the dataset formats and features we support.  The complication is that users really do expect Dask to support the 4th option (an arbitrary list of files).  The other challenge is that Dask's ability to support various partitioning schemes is completely dependent on the backend (which has historically been a bit of a moving target).

I do not personally have experience with many different parquet data formats, but I can say that my preference from a maintenence perspective is to support simple hive-like partitioning schemes, but only to the extent that they are already supported through the engine backend (arrow/fastparquet).  That is, we should allow pyarrow and/or fastparquet to discover/use partitioning information, but should try to avoid implementing any Dask-specific bells and whistles on top of the basic backend support.  For example, Allowing the user to explicitly map hive/directory partitions to Dask partitions would clearly be a bell/whistle that is not directly supported by either backend (and would require custom, and possibly fragile, Dask logic).

With my concern about custom/fragile Dask logic in mind, I do still agree that the regex-based chunksize-aggregation option is a bit ugly (even if it does happen to be powerful).  Therefore, it is okay with me to allow aggregation within a specific partition column.

Reply:
@fjetter - good job zooming out to bring up the larger picture.  Delta Lake is the most relevant technology for efficiently reading Parquet files (it's complementary with Arrow/FastParquet).

Vanilla Parquet lakes are limited.  They are easily corrupted (don't support transactions), allow files with incompatible schemas to be added, don't provide concurrency support, etc.  Delta Lake adds a transaction log on top of the Parquet lake, which allows for additional features, including better file reading.

@rjzamora - To efficiently read the Parquet files with Delta Lake, you'd first query the transaction log and then be able to structure the reads efficiently.  The transaction log contains metadata about the files, so you'd be able to figure out what small files should get read into the same partition.  It also contains partitioning information to prevent you from expensive "glob" file listing operations that are extremely slow on cloud based object storage systems like S3.  The Delta transaction log is a lot more powerful than the metadata files @fjetter was referring to.

I think reading in vanilla Parquet lakes more efficiently is a good move, but the long term solution for giving users a full featured Parquet lake experience is Delta Lake (or something similar).  I'll create a separate issue to hash this out more thoroughly.

Reply:
>I think reading in vanilla Parquet lakes more efficiently is a good move, but the long term solution for giving users a full featured Parquet lake experience is Delta Lake (or something similar). I'll create a separate issue to hash this out more thoroughly.

Thanks for sharing your Delta-Lake knowledge here @MrPowers -  My background is HPC/supercomputing, so I have very little background with parquet lake systems.  I would be happy to engage in a discussion about designing a better API for the Parquet-lake experience.  Given the "Frankenstein" state of the current parquet API, however, I suspect it will be tricky to work full-featured support into the existing design.

Reply:
> Delta Lake is the most relevant technology for efficiently reading Parquet files 

I do not necessarily agree with this statement but I agree that adding support for it would be a great addition. If you are experienced in this space, I would like to encourage you to open a dedicated ticket about introducing support for this since this clearly blows up the scope of this PR. However, expert knowledge would be highly appreciated. In particular if there are already existing libraries we can hook into to query existing delta lakes and write back to them.


---

> (4) an arbitrary list of files. I completely agree that it is much easier to design and maintain an API if we are explicit about the dataset formats and features we support. The complication is that users really do expect Dask to support the 4th option (an arbitrary list of files). 

I know this is kind of expected but we can (imho should) be picky in what kind of features we want support while living in the wild, wild west. Some features simply only make sense if there is already some structure to work with. I'm very certain that we cannot shoot for optimal performance if there is not structure to begin with.